src/terminal/adapter/MacroBuffer.cpp

```diff
@@ -0,0 +1,271 @@
+        {
+            _parseState = State::ExpectingRepeatCount;
+            _repeatCount = 0;
+        }
+        else if (ch == L';' && _repeatPending)
+        {
+            success = _applyPendingRepeat();
+        }
+        else
+        {
+            success = false;
+        }
+        break;
+    case State::ExpectingSecondHexDigit:
+        success = _decodeHexDigit(ch) && _appendToActiveMacro(_decodedChar);
+        _decodedChar = 0;
+        _parseState = State::ExpectingHexDigit;
+        break;
+    case State::ExpectingRepeatCount:
+        if (ch >= L'0' && ch <= L'9')
+        {
+            _repeatCount = _repeatCount * 10 + (ch - L'0');
+            _repeatCount = std::min<size_t>(_repeatCount, MAX_PARAMETER_VALUE);
+        }
+        else if (ch == L';')
+        {
+            _repeatPending = true;
+            _repeatStart = _activeMacro().length();
+            _parseState = State::ExpectingHexDigit;
+        }
+        else
+        {
+            success = false;
+        }
+        break;
+    default:
+        success = false;
+        break;
+    }
+
+    // If there is an error in the definition, clear everything received so far.
+    if (!success)
+    {
+        _deleteMacro(_activeMacro());
+    }
+    return success;
+}
+
+bool MacroBuffer::_decodeHexDigit(const wchar_t ch) noexcept
+{
+    _decodedChar <<= 4;
+    if (ch >= L'0' && ch <= L'9')
+    {
+        _decodedChar += (ch - L'0');
+        return true;
+    }
+    else if (ch >= L'A' && ch <= L'F')
+    {
+        _decodedChar += (ch - L'A' + 10);
+        return true;
+    }
+    else if (ch >= L'a' && ch <= L'f')
+    {
+        _decodedChar += (ch - L'a' + 10);
+        return true;
+    }
+    return false;
+}
+
+bool MacroBuffer::_appendToActiveMacro(const wchar_t ch)
+{
+    if (GetSpaceAvailable() > 0)
+    {
+        _activeMacro().push_back(ch);
+        _spaceUsed++;
+        return true;
+    }
+    return false;
+}
+
+std::wstring& MacroBuffer::_activeMacro() noexcept
+{
+    return til::at(_macros, _activeMacroId);
+}
+
+void MacroBuffer::_deleteMacro(std::wstring& macro) noexcept
+{
+    _spaceUsed -= macro.length();
+    std::wstring{}.swap(macro);
+}
+
+bool MacroBuffer::_applyPendingRepeat()
+{
+    auto& activeMacro = _activeMacro();
+    const auto sequenceLength = activeMacro.length() - _repeatStart;
+    if (sequenceLength && _repeatCount > 1)
+    {
+        // Note that the repeat sequence has already been written to the buffer
+        // once while it was being parsed, so we only need to append additional
+        // copies for repeat counts that are greater than one. If there is not
+        // enough space for the additional content, we'll just abort the macro.
+        const auto spaceRequired = (_repeatCount - 1) * sequenceLength;
+        if (spaceRequired > GetSpaceAvailable())
+        {
+            return false;
+        }
+        for (auto i = 1u; i < _repeatCount; i++)

```

Comment:
I believe this should use `size_t i = 1;` explicitly to get the proper 32/64-bit size.

Reply:
Yeah, OK. I wish we just had a size_t literal suffix, but I guess that'll have to wait until C++23.